source/fuzz/fuzzer_pass_replace_irrelevant_ids.cpp

```diff
@@ -0,0 +1,161 @@
+  for (auto irrelevant_id : irrelevant_ids) {
+    uint32_t type_id =
+        GetIRContext()->get_def_use_mgr()->GetDef(irrelevant_id)->type_id();
+
+    GetIRContext()->get_def_use_mgr()->ForEachUse(
+        irrelevant_id, [this, &irrelevant_id, &type_id, &types_to_ids,
+                        &transformations_to_apply](opt::Instruction* use_inst,
+                                                   uint32_t use_index) {
+          // The id must be used as an input operand.
+          if (use_index <
+              use_inst->NumOperands() - use_inst->NumInOperands()) {
+            // The id is used as an output operand, so we cannot replace this
+            // usage.
+            return;
+          }
+
+          // Randomly decide whether to consider this use.
+          if (!GetFuzzerContext()->ChoosePercentage(
+                  GetFuzzerContext()->GetChanceOfReplacingIrrelevantId())) {
+            return;
+          }
+
+          // Get the input operand index for this use, from the absolute operand
+          // index.
+          uint32_t in_index =
+              fuzzerutil::InOperandIndexFromOperandIndex(*use_inst, use_index);
+
+          // Only go ahead if this id use can be replaced in principle.
+          if (!fuzzerutil::IdUseCanBeReplaced(GetIRContext(), use_inst,
+                                              in_index)) {
+            return;
+          }
+
+          // Find out which ids could be used to replace this use.
+          std::vector<uint32_t> available_replacement_ids;
+
+          for (auto replacement_id : types_to_ids[type_id]) {
+            // We cannot replace an id with itself.
+            if (replacement_id == irrelevant_id) {
+              continue;
+            }
+
+            // Only consider this replacement if it is available at the id use
+            // point.
+            if (fuzzerutil::IdIsAvailableAtUse(GetIRContext(), use_inst,
+                                               in_index, replacement_id)) {
+              available_replacement_ids.push_back(replacement_id);
+            }
+          }
+
+          // Only go ahead if there is at least one id with which this use can
+          // be replaced.
+          if (available_replacement_ids.empty()) {
+            return;
+          }
+
+          // Choose the replacement id randomly.
+          uint32_t replacement_id =
+              available_replacement_ids[GetFuzzerContext()->RandomIndex(
+                  available_replacement_ids)];
+
+          // Add this replacement to the list of transformations to apply.
+          transformations_to_apply.emplace_back(
+              TransformationReplaceIrrelevantId(
+                  MakeIdUseDescriptorFromUse(GetIRContext(), use_inst,
+                                             in_index),
+                  replacement_id));
+        });
+  }
+
+  // Apply all the transformations.
+  for (auto transformation : transformations_to_apply) {
```

Comment:
Consider using `const auto&` to remove the needless call to the copy constructor.

Reply:
Done